evm/utils/test_builder/normalization.py

```diff
@@ -0,0 +1,278 @@
+from collections.abc import (
+    Iterable,
+    Mapping,
+)
+import functools
+
+from cytoolz import (
+    assoc_in,
+    compose,
+    concat,
+    identity,
+)
+from eth_utils import (
+    apply_formatters_to_dict,
+    big_endian_to_int,
+    decode_hex,
+    is_0x_prefixed,
+    is_bytes,
+    is_hex,
+    to_canonical_address,
+)
+import cytoolz.curried
+import eth_utils.curried
+
+from .builder_utils import (
+    deep_merge,
+    is_cleanly_mergable,
+)
+
+
+def state_definition_to_dict(state_definition):
+    """Convert a state definition to the canonical dict form.
+
+    State can either be defined in the canonical form, or as a list of sub states that are then
+    merged to one. Sub states can either be given as dictionaries themselves, or as tuples where
+    the last element is the value and all others the keys for this value in the nested state
+    dictionary. Example:
+
+    ```
+        [
+            ("0xaabb", "balance", 3),
+            ("0xaabb", "storage", {
+                4: 5,
+            }),
+            "0xbbcc", {
+                "balance": 6,
+                "nonce": 7
+            }
+        ]
+    ```
+    """
+    if isinstance(state_definition, Mapping):
+        state_dict = state_definition
+    elif isinstance(state_definition, Iterable):
+        state_dicts = [
+            assoc_in(
+                {},
+                state_item[:-1],
+                state_item[-1]
+            ) if not isinstance(state_item, Mapping) else state_item
+            for state_item
+            in state_definition
+        ]
+        assert is_cleanly_mergable(*state_dicts)
+        state_dict = deep_merge(*state_dicts)
+    else:
+        assert False
+
+    second_layer_keys = concat(d.keys() for d in state_dict.values())
+    assert all(key in ["balance", "nonce", "storage", "code"] for key in second_layer_keys)
```

Comment:
*minor performance thing*

- casting `second_layer_keys` to a `set` will remove duplicates so you don't check the same key multiple times.
- the set of allowed keys *should* be a `set` since we're using it purely to check for presence within the set.
- this could be done in one go with `bad_keys = set(seen_keys).difference({"balance", "nonce", ...})`

I'd recommend changing to use the `bad_keys = ` pattern because then you can raise an explicit error showing what *extra* keys were found.